### PR TITLE
PyTorch lightning based implementation for FuseMedML (generic and custom approach) 

### PR DESCRIPTION
**Pull request by mosheraboh**
Completed, you are welcome to review.

### CODE REVIEW

File: fuse/dl/managers/manager_default.py [hunk 1]
```diff
@@ -283,7 +283,9 @@ def should_load(object_name):
             checkpoint_objs.append(Checkpoint.load_from_file(checkpoint_file))
 
         if should_load('net'):
+            self.logger.info(f'extract checkpoints')
             net_state_dict_list = [checkpoint.net_state_dict for checkpoint in checkpoint_objs]
+            self.logger.info(f'load intonet')
```
Comment (mosheraboh) [inline fuse/dl/managers/manager_default.py:288]: remove

File: fuse/data/utils/split.py [hunk 2]
```diff
@@ -138,10 +137,9 @@ def dataset_balanced_division_to_folds(dataset: DatasetBase, output_split_filena
             keys = [get_sample_id_key(), id]
         if keys_to_balance is not None:
             keys += keys_to_balance
-        df = ExportDataset.export_to_dataframe(dataset, keys, workers = workers, mp_context=mp_context)
-        df_folds = balanced_division(df, id, keys_to_balance, nfolds, **kwargs)
+        df = ExportDataset.export_to_dataframe(dataset, keys)
```
Comment (mosheraboh) [inline fuse/data/utils/split.py:140]: revert

File: fuse/data/datasets/caching/samples_cacher.py [hunk 3]
```diff
@@ -342,7 +342,7 @@ def _cache(self, orig_sample_id:Any):
     
         
 
-def _get_available_write_location(cache_dirs:List[str], max_allowed_used_space=0.95):               
+def _get_available_write_location(cache_dirs:List[str], max_allowed_used_space=None):
```
Comment (mosheraboh) [inline fuse/data/datasets/caching/samples_cacher.py:345]: revert

File: examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+from fuse_examples.imaging.classification.mnist import lenet
+###########################################################################################################
+# Fuse
+###########################################################################################################
+
+##########################################
+# Lightning Module
+##########################################
+class LightningModuleMnist(pl.LightningModule):
+    def __init__(self, model_dir: str, opt_lr: float, opt_weight_decay: float, **kwargs):
```
Comment (alex-golts) [inline examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py:64]: please add some documentation

File: examples/fuse_examples/tests/test_classification_stoic21.py [hunk 5]
```diff
@@ -34,7 +34,7 @@
 class ClassificationStoic21TestCase(unittest.TestCase):
 
     def setUp(self):
-        self.root = tempfile.mkdtemp()
+        self.root = "tests2"#tempfile.mkdtemp()
```
Comment (alex-golts) [inline examples/fuse_examples/tests/test_classification_stoic21.py:37]: was this something temporary that you forgot to change back?

File: fuse/data/utils/split.py (hunk 2 above)
Comment (alex-golts) [reply fuse/data/utils/split.py:140]: what was the reason for these reverts?

**Review by alex-golts — COMMENTED** [review]
Looks really good! 
I wrote a few small comments inline

File: fuse/data/utils/split.py (hunk 2 above)
Comment (mosheraboh) [reply fuse/data/utils/split.py:140]: I meant that I should revert my modifications.
So you can ignore this one.

File: examples/fuse_examples/imaging/classification/mnist/run_mnist.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+from fuse.dl.losses.loss_default import LossDefault
+from fuse.dl.models.model_wrapper import ModelWrapSeqToDict
+from fuse.dl.lightning.pl_module import LightningModuleDefault
+
+from fuse.utils.utils_debug import FuseDebug
+from fuse.utils.utils_logger import fuse_logger_start
+
+from fuseimg.datasets.mnist import MNIST
+
+from fuse_examples.imaging.classification.mnist import lenet
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/mnist/run_mnist.py:54]: Thanks

File: examples/fuse_examples/imaging/classification/mnist/run_mnist.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+    # metrics
+    train_metrics = OrderedDict([
+        ('operation_point', MetricApplyThresholds(pred='model.output.classification')), # will apply argmax
+        ('accuracy', MetricAccuracy(pred='results:metrics.operation_point.cls_pred', target='data.label'))
+    ])
+
+    validation_metrics = copy.deepcopy(train_metrics) # use the same metrics in validation as well
+ 
+    # arguments to
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/mnist/run_mnist.py:169]: Stopped in the middle - will complete it, thanks.

File: examples/fuse_examples/imaging/classification/mnist/run_mnist.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+# Data
+# ============
+TRAIN_COMMON_PARAMS['data.batch_size'] = 100
+TRAIN_COMMON_PARAMS['data.train_num_workers'] = 8
+TRAIN_COMMON_PARAMS['data.validation_num_workers'] = 8
+
+# ===============
+# PL Trainer
+# ===============
+TRAIN_COMMON_PARAMS['trainer.num_epochs'] = 2
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/mnist/run_mnist.py:89]: Yes, it's about the same.
The results are already good after two epochs.

File: examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py (hunk 4 above)
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py:64]: Will do

File: examples/fuse_examples/tests/test_classification_stoic21.py (hunk 5 above)
Comment (mosheraboh) [reply examples/fuse_examples/tests/test_classification_stoic21.py:37]: temporary, thanks.

File: examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+# Fuse
+###########################################################################################################
+
+##########################################
+# Lightning Module
+##########################################
+class LightningModuleMnist(pl.LightningModule):
+    """
+    Implementation of pl.LightningModule
+    Demonstrate how to use FuseMedML with with your own PyTorch lightining implementaiton.
```
Comment (YoelShoshan) [inline examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py:66]: *with with

File: examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+        step_metrics(self._validation_metrics, batch_dict)
+
+        # return just the losses and drop everything else
+        return {"losses": batch_dict["losses"]} 
+     
+    def predict_step(self, batch_dict: NDict, batch_idx: int) -> dict:
+        # run forward function and store the outputs in batch_dict["model"]
+        batch_dict["model"] = self.forward(batch_dict)
+        # extract the requried keys - defined in self.set_predictions_keys()
+        return step_extract_predictions(self._prediction_keys, batch_dict)
```
Comment (YoelShoshan) [inline examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py:138]: perhaps an error if self._prediction_keys attribute is not found?
with instruction to call set_predictions_keys()

File: examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+from fuse.eval.evaluator import EvaluatorDefault 
+from fuse.eval.metrics.classification.metrics_thresholding_common import MetricApplyThresholds
+from fuse.eval.metrics.classification.metrics_classification_common import MetricAccuracy, MetricAUCROC, MetricROCCurve
+
+from fuse.data.utils.samplers import BatchSamplerDefault
+from fuse.data.utils.collates import CollateDefault
+
+from fuse.dl.losses.loss_default import LossDefault
+from fuse.dl.models.model_wrapper import ModelWrapSeqToDict
+from fuse.dl.lightning.pl_funcs import *
```
Comment (YoelShoshan) [inline examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py:48]: maybe instead it's better to do:
```import fuse.dl.lightning.pl_funcs as pl_funcs```
and then do pl_funcs.blah
pl_funcs.blah2
in the code?

**Review by YoelShoshan — DISMISSED** [review]
I think it looks great!
Few comments inline.

**Review by alex-golts — APPROVED** [review]

**Comment by mosheraboh** [comment]
Hi all,
I'm merging the PR to fuse2 branch.
You are still welcome to add comments here and I'll address them in a new PR.

**Pull request merged**